### ISSUE TITLE
TF_WORKSPACE must be explicitly defined in tags, rather than matching on the configured tags

### ISSUE THREAD

**Issue by robwittman** [opentofu/opentofu#1787]
### OpenTofu Version

```shell
1.7.3
```

### OpenTofu Configuration Files

```hcl
terraform {
  cloud {
    organization = "$organizationId"
    hostname = "hostname.com"

    workspaces {
      tags = ["demo"]
    }
  }
}

variable "region" {
  type = string
}

resource "random_string" "test" {
  length = 8
}

output "region_suffix" {
  value = "${var.region}-${random_string.test.result}"  
}
```

### Debug Output

```
TF_LOG=trace TF_WORKSPACE=tofu-demo-1 tofu plan
2024-07-09T12:09:36.910-0400 [INFO]  OpenTofu version: 1.7.3
2024-07-09T12:09:36.910-0400 [DEBUG] using github.com/hashicorp/go-tfe v1.36.0
2024-07-09T12:09:36.910-0400 [DEBUG] using github.com/opentofu/hcl/v2 v2.0.0-20240416130056-03228b26f391
2024-07-09T12:09:36.910-0400 [DEBUG] using github.com/hashicorp/terraform-svchost v0.1.1
2024-07-09T12:09:36.910-0400 [DEBUG] using github.com/zclconf/go-cty v1.14.4
2024-07-09T12:09:36.910-0400 [INFO]  Go runtime version: go1.21.3
2024-07-09T12:09:36.910-0400 [INFO]  CLI args: []string{"/Users/rwittman/.asdf/installs/opentofu/1.7.3/bin/tofu", "plan"}
2024-07-09T12:09:36.910-0400 [TRACE] Stdout is a terminal of width 108
2024-07-09T12:09:36.910-0400 [TRACE] Stderr is a terminal of width 108
2024-07-09T12:09:36.910-0400 [TRACE] Stdin is a terminal
2024-07-09T12:09:36.910-0400 [DEBUG] Attempting to open CLI config file: /Users/rwittman/.tofurc
2024-07-09T12:09:36.910-0400 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2024-07-09T12:09:36.910-0400 [INFO]  Loading CLI configuration from /Users/rwittman/.terraform.d/credentials.tfrc.json
2024-07-09T12:09:36.911-0400 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2024-07-09T12:09:36.911-0400 [DEBUG] ignoring non-existing provider search directory /Users/rwittman/.terraform.d/plugins
2024-07-09T12:09:36.911-0400 [DEBUG] ignoring non-existing provider search directory /Users/rwittman/Library/Application Support/io.terraform/plugins
2024-07-09T12:09:36.911-0400 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2024-07-09T12:09:36.911-0400 [DEBUG] Found the config directory: /Users/rwittman/.terraform.d
2024-07-09T12:09:36.911-0400 [INFO]  CLI command args: []string{"plan"}
2024-07-09T12:09:36.911-0400 [TRACE] Meta.Backend: built configuration for "cloud" backend with hash value 2856480487
2024-07-09T12:09:36.911-0400 [TRACE] Preserving existing state lineage "5062a514-a077-340b-638b-033b9d5fcc1c"
2024-07-09T12:09:36.911-0400 [TRACE] Preserving existing state lineage "5062a514-a077-340b-638b-033b9d5fcc1c"
2024-07-09T12:09:36.911-0400 [TRACE] Meta.Backend: working directory was previously initialized for "cloud" backend
2024-07-09T12:09:36.911-0400 [TRACE] Meta.Backend: using already-initialized, unchanged "cloud" backend configuration
╷
│ Error: Invalid workspaces configuration
│ 
│ The workspace defined using the environment variable "TF_WORKSPACE" does not belong to "tags".
│ 
│ The 'workspaces' block configures how OpenTofu CLI maps its workspaces for this single
│ configuration to workspaces within a cloud backend organization. Two strategies are available:
│ 
│ [bold]tags[reset] - A set of tags used to select remote cloud backend workspaces to be used for this
│ single
│ configuration. New workspaces will automatically be tagged with these tag values. Generally, this
│ is the primary and recommended strategy to use. This option conflicts with "name".
│ 
│ [bold]name[reset] - The name of a single cloud backend workspace to be used with this configuration.
│ When configured, only the specified workspace can be used. This option conflicts with "tags"
```

### Expected Behavior

I have an implementation of the cloud backend for usage with OpenTofu. If I have 2 workspaces, with the appropriate tags, such as 
```
tofu-demo-1: {"tags":["demo"]}
tofu-demo-2: { "tags": ["demo"]}
```
I expect that I should be able to do 
```shell 
tofu init 
TF_WORKSPACE=tofu-demo-1 tofu plan
TF_WORKSPACE=tofu-demo-2 tofu plan
```
to operate on 2 different workspaces in the cloud, while using the same configuration. 

### Actual Behavior

Tofu seems to require that the workspace defined by`TF_WORKSPACE` be defined explicitly in the list of tags, which I think defeats the purpose of using tag based selection

### Steps to Reproduce

This would be fairly hard to reproduce, as it requires a server, or at least static endpoints, that satisfies the cloud backend. But with that in place, I should be able to do 
```
tofu init 
TF_WORKSPACE=terraform-demo-1 tofu plan
TF_WORKSPACE=terraform-demo-2 tofu plan
```

to run plans on 2 separate cloud workspaces, without having to change the underlying configuration

### Additional Context

I've been working on an open source application that satisfies the API endpoints required for the `cloud` backend. Most of the related operations are standard OpenTofu usage.

### References

- https://github.com/opentofu/opentofu/pull/867/files

I believe this is the code causing the error state. Ideally, this wouldn't check if the workspace name is in the list of tags, but rather fetch the workspace information from the API, and verify that the configured tags match the tags on the workspace
- https://github.com/opentofu/opentofu/blob/main/internal/cloud/backend.go#L470

**Comment by cam72cam** [comment]
Hey, thanks for the bug report!  We'll discuss this a bit more during triage this week and prioritize accordingly.  It looks like there's some potential spots of confusion hinted at in the PR you linked to: https://github.com/opentofu/opentofu/pull/867/files#r1394147328

**Comment by Evi1Pumpkin** [comment]
Hey @robwittman,
Thanks for reporting! I investigated and believe you are correct. We chose the incorrect solution in the PR you mentioned because of the misleading wording in the documentation.
I'll open a PR to fix it shortly. Unfortunately, I have some issues testing it end to end because none of the `cloud` backends I can access support the workspaces `tags` attribute. You said you were working on an Open-Source backend. Would it be possible for me to access it so that I can test the fix? Another option is that I'll implement the fix in a PR and ask you to test it to see if it works correctly.
What do you think? Does that sound reasonable to you?

**Comment by robwittman** [comment]
@Evi1Pumpkin Happy to share the repo and some instructions privately (not quite ready to be public with it yet 😅 ), or I can share a URL to one of my environments. 

I'm also happy to test any PRs you have for it against my local instance.

**Comment by robwittman** [comment]
@Evi1Pumpkin Eh, no better time than the present haha

Here's a [gist](https://gist.github.com/robwittman/64312f3ece7a13d01198d9d10f8723ed) for cloning the project and doing some basic setup. The UI is not great at the moment, but most of what needs testing is API / CLI anyways. 

If you have any issues or feedback on setting it up, let me know. I tore down my sandbox environment so there's nothing I can grant access to at the moment, so I can also do some test verification on my end if it proves troublesome

**Comment by Evi1Pumpkin** [comment]
@robwittman, posting your implementation of your backend was super helpful! I managed to spin it up quite easily (with one tweak of replacing the hardcoded Ngrok addressed in the code). After that, I could test my implementation, and I realized my original implementation had a bug. I fixed that and posted [a PR](https://github.com/opentofu/opentofu/pull/1930).
Thank you so much for sharing your cloud backend!
